src/Illuminate/Http/Resources/CollectsResources.php

```diff
@@ -45,15 +47,21 @@ protected function collectResource($resource)
      */
     protected function collects()
     {
-        if ($this->collects) {
-            return $this->collects;
-        }
+        $collects = null;
 
-        if (str_ends_with(class_basename($this), 'Collection') &&
+        if ($this->collects) {
+            $collects = $this->collects;
+        } elseif (str_ends_with(class_basename($this), 'Collection') &&
             (class_exists($class = Str::replaceLast('Collection', '', get_class($this))) ||
              class_exists($class = Str::replaceLast('Collection', 'Resource', get_class($this))))) {
-            return $class;
+            $collects = $class;
         }
+
+        if (! $collects || is_subclass_of($collects, JsonResource::class)) {
+            return $collects;
+        }
+
+        throw new RuntimeException('ResourceCollection must collect JsonResources.');
```

Comment:
Wouldn't it make more sense to throw a `InvalidArgumentException` here?

Reply:
@axlon it's not really an argument for this method that's used, more like a preset value on the object's property. In this case it can be a LogicException: https://www.php.net/manual/en/class.logicexception.php as the dev is able to fix this directly in their code.

Reply:
Yeah I was debating between the 2, but you're right `LogicException` makes more sense 👍

Reply:
@bastien-phi can you adjust that?

Reply:
done !